orca-keel/src/main/kotlin/com/netflix/spinnaker/orca/keel/task/UpsertIntentTask.kt

```diff
@@ -0,0 +1,91 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.keel.task
+
+import com.fasterxml.jackson.core.type.TypeReference
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.netflix.spinnaker.orca.ExecutionStatus
+import com.netflix.spinnaker.orca.KeelService
+import com.netflix.spinnaker.orca.RetryableTask
+import com.netflix.spinnaker.orca.TaskResult
+import com.netflix.spinnaker.orca.keel.model.Intent
+import com.netflix.spinnaker.orca.keel.model.UpsertIntentDryRunResponse
+import com.netflix.spinnaker.orca.keel.model.UpsertIntentRequest
+import com.netflix.spinnaker.orca.keel.model.UpsertIntentResponse
+import com.netflix.spinnaker.orca.pipeline.model.Stage
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.http.HttpStatus
+import org.springframework.stereotype.Component
+import java.util.HashMap
+import java.util.concurrent.TimeUnit
+
+@Component
+class UpsertIntentTask
+@Autowired constructor(
+  private val keelService: KeelService,
+  private val keelObjectMapper: ObjectMapper
+) : RetryableTask {
+  private val log = LoggerFactory.getLogger(javaClass)
+
+  override fun execute(stage: Stage) : TaskResult {
+    val missingParams = mutableListOf<String>()
+
+    if (!stage.context.containsKey("intents")) {
+      missingParams.add("intents")
+    }
+
+    if (!stage.context.containsKey("dryRun") ) {
+      missingParams.add("dryRun")
+    }
+
+    if (missingParams.isNotEmpty()) {
+      throw IllegalArgumentException("Missing required task parameters (${missingParams.joinToString(",")})")
+    }
+
+    val upsertIntentRequest = UpsertIntentRequest(
+      intents = keelObjectMapper.convertValue(stage.context["intents"], object : TypeReference<List<Intent>>() {}),
+      dryRun = if (stage.context["dryRun"] == null) false else stage.context["dryRun"].toString().toBoolean()
+    )
+
+    val response = keelService.upsertIntents(upsertIntentRequest)
+
+    val outputs = HashMap<String, Any>()
+
+    try {
+      if (upsertIntentRequest.dryRun) {
+        val dryRunResponse = keelObjectMapper.readValue<List<UpsertIntentDryRunResponse>>(response.body.`in`(), object : TypeReference<List<UpsertIntentDryRunResponse>>(){})
+        outputs.put("upsertIntentResponse", dryRunResponse)
+      } else {
+        val upsertResponse = keelObjectMapper.readValue<List<UpsertIntentResponse>>(response.body.`in`(), object : TypeReference<List<UpsertIntentResponse>>(){})
+        outputs.put("upsertIntentResponse", upsertResponse)
+      }
+    } catch (e: Exception) {
+        throw RuntimeException("Error processing upsert intent response from keel", e)
+    }
```

Comment:
Is there a better exception type? Could we just not catch?

Reply:
An error here indicates that we can't understand the response from keel. It might not be a big deal, but for now I want to pass it along I think? I'll throw a better exception though.

Reply:
Actually maybe I"ll just log it for now. Doesn't seem worth stopping everything for.

Reply:
Yeah, I'd log & re-throw